src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -703,51 +808,42 @@ private async Task SendDataAsync(ReadOnlyMemory<byte> buffer, CancellationToken
             {
                 ReadOnlyMemory<byte> remaining = buffer;
 
-                while (remaining.Length > 0)
+                // Deal with ActiveIssue #9071:
+                // Custom HttpContent classes do not get passed the cancellationToken.
+                // So, inject the expected CancellationToken here, to ensure we can cancel the request body send if needed.
+                CancellationTokenSource customCancellationSource = null;
+                if (!cancellationToken.CanBeCanceled)
                 {
-                    int sendSize = await _streamWindow.RequestCreditAsync(remaining.Length, cancellationToken).ConfigureAwait(false);
-
-                    ReadOnlyMemory<byte> current;
-                    (current, remaining) = SplitBuffer(remaining, sendSize);
-
-                    await _connection.SendStreamDataAsync(_streamId, current, cancellationToken).ConfigureAwait(false);
+                    cancellationToken = _requestBodyCancellationSource.Token;
+                }
+                else if (cancellationToken != _requestBodyCancellationSource.Token)
+                {
+                    // User passed a custom CancellationToken.
+                    // We can't tell if it includes our Token or not, so assume it doesn't.
+                    customCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _requestBodyCancellationSource.Token);
```

Comment:
Doesn't this mean that all of our built-in HttpContent types, e.g. StreamContent, are going to allocate a new CancellationTokenSource for every write?

Reply:
Or, in that case cancellationToken will == _requestBodyCancellationSource.Token I guess?

Reply:
> Or, in that case cancellationToken will == _requestBodyCancellationSource.Token I guess?

That's the idea, yeah.